File: src/Lesson4/java/Ex2.java
package Lesson4.java;

import java.util.Scanner;

public class Ex2 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter string:");
        String input = scanner.nextLine();
        String reversed = reverseWords(input);
        System.out.println("Result: " + reversed);
    }

    public static String reverseWords(String str) {
        StringBuilder result = new StringBuilder();
        int i = str.length() - 1;

        while (i >= 0) {

            int end = i;
            while (i >= 0 && str.charAt(i) != ' ') {
                i--;
            }

            String word = str.substring(i + 1, end + 1);
            result.append(word).append(" ");

            if (i >= 0) {
                i--;
            }
        }
        return result.toString().trim();
    }
}

File: src/Lesson4/forkotlin/Ex2.kt
package Lesson4.forkotlin

fun main() {
    print("Enter string: ")
    val input = readln()

    val reversed = reverseWords(input)
    println("Result: $reversed")
}

fun reverseWords(str: String): String {
    if (str.isBlank()) return ""

    val result = StringBuilder()
    var i = str.length - 1

    while (i >= 0) {

        while (i >= 0 && str[i] == ' ') {
            i--
        }

        if (i < 0) break

        val end = i

        while (i >= 0 && str[i] != ' ') {
            i--
        }
        val word = str.substring(i + 1, end + 1)

        if (result.isNotEmpty()) {
            result.append(" ")
        }
        result.append(word)
    }

    return result.toString()
}